src/main/java/org/icgc/argo/program_service/services/ego/EgoService.java

```diff
@@ -0,0 +1,253 @@
+import javax.validation.constraints.Email;
+import java.security.interfaces.RSAPublicKey;
+import java.util.*;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static org.icgc.argo.program_service.proto.UserRole.ADMIN;
+import static org.icgc.argo.program_service.services.ego.GroupName.createProgramGroupName;
+
+//TODO [rtisma]: refactor into service and client
+@Slf4j
+@Service
+public class EgoService {
+  private final EgoRESTClient egoClient;
+  private final ProgramEgoGroupRepository programEgoGroupRepository;
+  private final ProgramConverter programConverter;
+  private RSAPublicKey egoPublicKey;
+
+  @Autowired
+  public EgoService(
+    @NonNull ProgramEgoGroupRepository programEgoGroupRepository,
+	  @NonNull ProgramConverter programConverter,
+    @NonNull EgoRESTClient restClient
+      ) {
+    this.programEgoGroupRepository = programEgoGroupRepository;
+    this.programConverter = programConverter;
+    this.egoClient = restClient;
+  }
+
+  @Autowired
+  private void setEgoPublicKey() {
+    this.egoPublicKey = egoClient.getPublicKey();
+  }
+
+
+  public Optional<EgoToken> verifyToken(String jwtToken) {
+    try {
+      Algorithm algorithm = Algorithm.RSA256(this.egoPublicKey, null);
+      JWTVerifier verifier = JWT.require(algorithm)
+        .withIssuer("ego")
+        .build(); //Reusable verifier instance
+      val jwt = verifier.verify(jwtToken);
+      return parseToken(jwt);
+    } catch (JWTVerificationException | NullPointerException e) {
+      // Handle NPE defensively for null JWT.
+      return Optional.empty();
+    }
+  }
+
+  public Optional<EgoToken> parseToken(DecodedJWT jwt) {
+    try {
+      EgoToken egoToken = new EgoToken(jwt, jwt.getClaim("context").as(Context.class));
+      return Optional.of(egoToken);
+    } catch (JWTDecodeException exception){
+      //Invalid token
+      return Optional.empty();
+    }
+  }
+
+
+  //TODO: add transactional. If there are more programdb logic in the future and something fails, it will be able to roll back those changes.
+  public void setUpProgram(@NonNull ProgramEntity program, @NonNull Collection<String> adminEmails) {
+    val programPolicy = egoClient.createEgoPolicy("PROGRAM-" + program.getShortName());
+    val dataPolicy = egoClient.createEgoPolicy("PROGRAMDATA-" + program.getShortName());
+
+    Stream.of(UserRole.values())
+      .filter(role -> !role.equals(UserRole.UNRECOGNIZED))
+      .forEach(
+        role -> {
+          val group = ensureGroupExists(program, role);
+          egoClient.assignPermission(group, programPolicy, getProgramMask(role));
+          egoClient.assignPermission(group, dataPolicy, getDataMask(role));
+          saveGroupIdForProgramAndRole(program, role, group.getId());
+
+        }
+      );
+
+    adminEmails.forEach(email -> initAdmin(email, program));
+  }
+
+  public String getProgramMask(UserRole role) {
```

Comment:
any reason why we can't make this static? or private?

Reply:
make this private static since no state needed